linera-core/src/unit_tests/client_tests.rs

```diff
@@ -1811,7 +2341,131 @@ where
+    builder
+        .set_fault_type([1, 2], FaultType::DontSendConfirmVote)
+        .await;
+    builder.set_fault_type([3], FaultType::Offline).await;
+    let result = client0
+        .burn(None, Amount::from_tokens(3), UserData::default())
+        .await;
+    assert!(result.is_err());
+    let manager = client0
+        .chain_info_with_manager_values()
+        .await
+        .unwrap()
+        .manager;
+    // Validator 0 may or may not have processed the validated block before the update was
+    // canceled due to the errors from the faulty validators. Submit it again to make sure
+    // it's there, so that client 1 can download and re-propose it later.
+    let validated_block_certificate = *manager.requested_locked.unwrap();
+    builder
+        .node(0)
+        .handle_certificate(
+            validated_block_certificate,
+            Vec::new(),
+            Vec::new(),
+            CrossChainMessageDelivery::Blocking,
+        )
+        .await
+        .unwrap();
+
+    // Client 1 wants to burn 2 tokens. They learn about the proposal in round 0, but now the
+    // validator 0 is offline, so they don't learn about the validated block and make their own
+    // proposal in round 1.
+    builder.set_fault_type([0], FaultType::Offline).await;
+    builder
+        .set_fault_type([3], FaultType::OfflineWithInfo)
+        .await;
+    client1.synchronize_from_validators().await.unwrap();
+    let manager = client1
+        .chain_info_with_manager_values()
+        .await
+        .unwrap()
+        .manager;
+    assert!(manager.requested_proposed.is_some());
+    assert!(manager.requested_locked.is_some());
+    assert_eq!(manager.current_round, Round::MultiLeader(0));
+    let result = client1
+        .burn(None, Amount::from_tokens(2), UserData::default())
+        .await;
+    assert!(result.is_err());
+
+    // Finally, three validators are online and honest again. Client 1 realizes there has been a
+    // validated block in round 0, and re-proposes it when it tries to burn 4 tokens.
+    builder.set_fault_type([0, 1, 2], FaultType::Honest).await;
+    builder.set_fault_type([3], FaultType::Offline).await;
+    client1.synchronize_from_validators().await.unwrap();
+    let manager = client1
+        .chain_info_with_manager_values()
+        .await
+        .unwrap()
+        .manager;
+    assert_eq!(
+        manager.requested_locked.unwrap().round,
+        Round::MultiLeader(0)
+    );
+    assert_eq!(manager.current_round, Round::MultiLeader(0));
+    assert!(client1.pending_block().is_none());
+    client1
+        .burn(None, Amount::from_tokens(4), UserData::default())
+        .await
+        .unwrap();
+
+    // Burning 3 and 4 tokens got finalized; the pending 2 tokens got skipped.
+    client0.synchronize_from_validators().await.unwrap();
+    assert_eq!(
+        client0.local_balance().await.unwrap(),
+        Amount::from_tokens(3)
+    );
+    Ok(())
+}
```

Comment:
Graphite is showing this whole test as new—is it? How does it differ from `test_re_propose_validated2`, and why isn't it redundant because of `test_re_propose_locked_block_with_blobs`? Sorry for the confusion, but these are too many _very_ similar tests. Also, we should be consistent in the naming and either call all of them `re_propose_locked` or `re_propose_validated`. (My bad.)